clubs/mixins.py

```diff
@@ -0,0 +1,241 @@
+        return str(value)
+
+    def _many_to_many_formatter(self, value):
+        """
+        Take in output generated by a ManyToMany serializer and format it.
+        """
+        if isinstance(value, list):
+            return ', '.join(self._many_to_many_individual_formatter(v) for v in value)
+        return value
+
+    def _lookup_field_formatter(self, field):
+        """
+        Return a method that can format the field given the name of the field.
+        """
+        # if there is no model serializer, don't format
+        try:
+            serializer = self.get_serializer_class()
+        except AttributeError:
+            return lambda x: x
+
+        if not isinstance(serializer, serializers.SerializerMetaclass):
+            return lambda x: x
+
+        # allow serializer to override field formatting
+        key = 'format_{}_for_spreadsheet'.format(field)
+        if hasattr(serializer, key):
+            return lambda x: getattr(serializer, key)(serializer, x)
+
+        # lookup column type from serializer
+        if field in serializer._declared_fields:
+            serializer_field_object = serializer._declared_fields[field]
+            if isinstance(serializer_field_object, serializers.SerializerMethodField):
+                return lambda x: x
+
+        # lookup column type from model
+        model = serializer.Meta.model
+        try:
+            field_object = model._meta.get_field(field)
+        except FieldDoesNotExist:
+            # if model field lookup fails, rely on serializer field
+            if isinstance(serializer_field_object, serializers.BooleanField):
+                return lambda x: str(bool(x))
+            return lambda x: x
+
+        # format based on field type
+        if isinstance(field_object, ManyToManyField):
+            return self._many_to_many_formatter
+        elif isinstance(field_object, BooleanField):
+            return lambda x: str(bool(x))
+        elif hasattr(field_object, 'choices') and field_object.choices is not None:
+            choices = dict(field_object.choices)
+            return lambda x: choices.get(x, 'Unknown')
+        else:
+            return lambda x: x
+
+    def _format_cell(self, key, value):
+        """
+        Format a cell in the exported Excel spreadsheet given (column name, cell value).
+        Returns (new column name, new cell value). Looks up formatting information using
+        the modal obtained from the serializer.
+        """
+        new_key = self._format_header_value(key)
+        # cache column formatter by column name
+        if key not in self._field_dict:
+            self._field_dict[key] = self._lookup_field_formatter(key)
+        return (new_key, self._field_dict[key](value))
+
+    def get_filename(self):
+        """
+        Returns a custom filename for the spreadsheet.
+        """
+        return 'report-{}.xlsx'.format(datetime.datetime.now().strftime('%Y%m%d-%H%M'))
```

Comment:
I think selina's design allowed custom filenames to be set on the frontend. Not sure if this is still over-ridden in the view?

Reply:
This functionality was actually never implemented in the original pull request. I made this mixin as generic as possible (ideally we could put this in `drf-renderer-xlsx`, but I'm not confident that it handles all of the edge cases). This mixin would also not be able to access that filename variable without a ton of introspection.

We can override this in a later pull request in the actual ViewSet!

Reply:
oh hmm i guess i never actually implemented this haha. sounds good!